### PR TITLE
Add OpenSim::updateConnecteesBySearch().

### PR DESCRIPTION
**Pull request by chrisdembia**
Add utility function to fix conectee paths from intermediate model files (e.g., 30516).

I created this utility to update the jumper model from WCB2018, which was at version 30516, for use in the 4.0 webinar.

---
This change is [Reviewable](https://reviewable.io/reviews/opensim-org/opensim-core/2333)

### CODE REVIEW

**Comment by chrisdembia** [comment]
This should not be merged until after #2335 is merged.

File: OpenSim/Simulation/SimulationUtilities.h [hunk 1]
```diff
@@ -94,6 +94,22 @@ void updatePre40KinematicsFilesFor40MotionType(const Model& model,
         const std::vector<std::string>& filePaths,
         std::string suffix="_updated");
 
+/** This function attempts to update the connectee path for any Socket anywhere
+ * in the model whose connectee path does not point to an existing component.
+ * The paths are updated by searching the model for a component with the
+ * correct name. For example, a connectee path like
+ * `../../some/invalid/path/to/foo` will be updated to `/bodyset/foo` if a Body
+ * named `foo` exists in the Model's BodySet.
```
Comment (tkuchida) [inline OpenSim/Simulation/SimulationUtilities.h]: Specify what happens if there are two bodies named "foo"?

File: OpenSim/Simulation/SimulationUtilities.h [hunk 2]
```diff
@@ -94,6 +94,22 @@ void updatePre40KinematicsFilesFor40MotionType(const Model& model,
...
+ *
+ * This function does not update the connectee paths for Inputs.
+ *
+ * This function is mainly for use with models loaded from files with XML
```
Comment (tkuchida) [inline OpenSim/Simulation/SimulationUtilities.h]: "mainly for use with" doesn't really specify what I can/should do. Perhaps something like "This method is intended for use with models loaded from version-30516 XML files to bring them up to date with the 4.0 API". I would prefer to see a report after execution saying either "No Sockets needed updating; Model was not changed" or "x Sockets were updated".

File: OpenSim/Simulation/SimulationUtilities.h [hunk 3]
```diff
@@ -94,6 +94,22 @@ void updatePre40KinematicsFilesFor40MotionType(const Model& model,
...
         std::string suffix="_updated");
 
+/** This function attempts to update the connectee path for any Socket anywhere
+ * in the model whose connectee path does not point to an existing component.
+ * The paths are updated by searching the model for a component with the
+ * correct name. For example, a connectee path like
+ * `../../some/invalid/path/to/foo` will be updated to `/bodyset/foo` if a Body
+ * named `foo` exists in the Model's BodySet.
+ *
+ * This function does not update the connectee paths for Inputs.
```
Comment (tkuchida) [inline OpenSim/Simulation/SimulationUtilities.h]: Suggest removing this comment and renaming the method to `updateSocketConnecteesBySearch`.

File: OpenSim/Simulation/SimulationUtilities.cpp [hunk 4]
```diff
@@ -175,3 +175,32 @@ void OpenSim::updatePre40KinematicsFilesFor40MotionType(const Model& model,
...
+                socket.finalizeConnection(model);
+            } catch (const ComponentNotFoundOnSpecifiedPath&) {
+                const ComponentPath path(socket.getConnecteePath());
+                if (path.getNumPathLevels() >= 1) { 
+                    const Component* found =
+                        model.findComponent(path.getComponentName());
+                    if (found) {
+                        socket.connect(*found);
+                        socket.finalizeConnection(model);
+                    }
```
Comment (tkuchida) [inline OpenSim/Simulation/SimulationUtilities.cpp]: else, report that this Socket needs updating but couldn't be done automatically?

Comment (chrisdembia) [reply OpenSim/Simulation/SimulationUtilities.cpp]: Done.

File: OpenSim/Simulation/SimulationUtilities.h (hunk 3 above)
Comment (chrisdembia) [reply OpenSim/Simulation/SimulationUtilities.h]: Done.

File: OpenSim/Simulation/SimulationUtilities.h (hunk 2 above)
Comment (chrisdembia) [reply OpenSim/Simulation/SimulationUtilities.h]: Done.

File: OpenSim/Simulation/SimulationUtilities.h (hunk 1 above)
Comment (chrisdembia) [reply OpenSim/Simulation/SimulationUtilities.h]: Done.

**Comment by chrisdembia** [comment]
Thanks @tkuchida I addressed your comments.

**Comment by chrisdembia** [comment]
Thanks.

**Pull request merged**